### PR TITLE
Byte buf proposal

### PR DESCRIPTION
**Pull request by bretambrose**
Attempt 2

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: include/aws/crt/Outcome.h [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+
+            const R &GetResult() const { return m_result; }
+
+            R &GetResult() { return m_result; }
+
+            /**
+             * casts the underlying result to an r-value so that caller can take ownership of underlying resources.
+             * this is necessary when streams are involved.
+             */
+            R &&AcquireResult() { return std::move(m_result); }
```
Comment (graebm) [inline include/aws/crt/Outcome.h]: I think this helper function is more confusing than just letting users call `std::move(outcome.GetResult())`.

The standard library doesn't offer any helper functions like this. If you want to acquire the contents of an optional<T> you just do `std::move(optional.value())` or `std::move(*optional)`

File: include/aws/crt/Outcome.h [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+
+            const E &GetError() const { return m_error; }
+
+            operator bool() const { return m_success; }
+
+          private:
+            R m_result;
+            E m_error;
```
Comment (graebm) [inline include/aws/crt/Outcome.h]: problems for another day: Both R and E types need to be default-constructible for this to work. Like, how will this work with HttpConnection whose default-constructor will be private?

File: include/aws/crt/Outcome.h [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+    {
+        template <typename R, typename E> // Result, Error
+        class Outcome
+        {
+          public:
+            Outcome() : m_success(false) {}
+
+            Outcome(const R &r) : m_result(r), m_success(true) {}
+
+            Outcome(const E &e) : m_error(e), m_success(false) {}
```
Comment (graebm) [inline include/aws/crt/Outcome.h]: Problems for another day: Works until R and E might be confused for one another by the compiler.
Ex: `AwsCrtResult<int> HttpClientStream::GetStatusCode()`

**Review by graebm — COMMENTED** [review]
We don't need to address all these edge-cases. I'm just being a know-it-all who also wrote an Outcome class once.

File: include/aws/crt/Outcome.h (hunk 2 above)
Comment (bretambrose) [reply include/aws/crt/Outcome.h]: True, but in that case I would assume you'd be getting a pointer back, not an instance.

File: include/aws/crt/Outcome.h (hunk 1 above)
Comment (bretambrose) [reply include/aws/crt/Outcome.h]: Updated

File: include/aws/crt/Outcome.h (hunk 3 above)
Comment (bretambrose) [reply include/aws/crt/Outcome.h]: True, but I wish we had a specific error type other than int as well, which would make things much less likely.

**Comment by bretambrose** [comment]
On second thought, let's not.

**Pull request closed without merging**